### PR TITLE
Scroll to bottom of messages if user has seen thread before

### LINKED ISSUE #2658: Improve thread scrolling based on improved threadLastSeen logic
The current thread scrolling behavior is to push you to the bottom when you are either a participant or thread creator. Up until now this has been our best guess at if you should auto jump to see the latest messages.

But now that we know if a user has visited a thread before, we should incorporate that into the scroll logic, so that when I click to view a thread with the 'new messages' badge it will push me to the bottom to see the new messages!

### PR DESCRIPTION
**Pull request by brianlovin**
**Status**
- [ ] WIP
- [x] Ready for review
- [ ] Needs testing

**Deploy after merge (delete what needn't be deployed)**
- hyperion (frontend)

This fixes an annoying issue where you would always have to manually scroll to the bottom of a thread where you weren't a participant, but wanted to follow along with the conversation. Now we auto scroll to the bottom of the messages if you've seen that thread before.

Closes #2658

### CODE REVIEW

**Comment by brianlovin** [comment]
Not sure what's up with these CI failures

screenshot 2018-05-04 16 35 54

**Review by mxstbr — COMMENTED** [review]
This shouldn't scroll to the bottom it should scroll to the "New messages" indicator if there is one, otherwise to the bottom. Would that be much harder to do? That'd honestly be the _ideal_ experience.

**Comment by ryota-murakami** [comment]
probably @mxstbr mentioned following UI(`New messages`).

screen shot 2018-05-14 at 1 47 19

as far as i know the "New messages" indicator impl following section.
https://github.com/withspectrum/spectrum/blob/alpha/src/components/messageGroup/index.js#L200-L223

and i can see Scroll to bottom impl at https://github.com/withspectrum/spectrum/blob/alpha/src/views/directMessages/containers/existingThread.js#L76-L80

but i can't see Scroll to bottom impl into codebase, so you might have to create new function anywhere:grinning:

**Comment by brianlovin** [comment]
Can we ship this first and worry about scrolling to the 'last messages' indicator later? As it is, we're already handling the message fetching pagination settings to only fetch unseen messages anyways, so it's likely that those will be close by.

File: src/views/thread/container.js [hunk 1]
```diff
@@ -325,9 +325,10 @@ class ThreadContainer extends React.Component<Props, State> {
         isCommunityOwner ||
         isChannelModerator ||
         isCommunityModerator;
-      const isParticipantOrAuthor =
```
Comment (mxstbr) [inline src/views/thread/container.js]: We do actually use the `isParticipantOrAuthor` var further down below (L518) so we can't just remove it—that's why CI fails.

**Review by mxstbr — CHANGES_REQUESTED** [review]

File: src/views/thread/container.js (hunk 1 above)
Comment (brianlovin) [reply src/views/thread/container.js]: Have an in-progress refactor locally, didn't mean to leave this PR marked as in-review :) thanks!

**Comment by brianlovin** [comment]
Okay @mxstbr https://github.com/withspectrum/spectrum/pull/3026/commits/07dda84c17455f15ca868989136c9e7a2fe19745 is my refactor - this should *definitely* be on alpha for a bit to test, but in my local testing I have fixed the problems of threads not autoscrolling if you navigate away from the inbox and then go back in.

If you have time to approve this today I can get it on alpha and test throughout the day.

**Comment by brianlovin** [comment]
Oh, some notes to add context to the refactor:

- in the thread container we used to check if the thread was a watercooler or not, and then return a component that was like 95% similar to a normal thread component. I just went ahead and made it so that last 5% is handled conditionally in the render and we don't have to maintain 2 almost-the-same thread components
- i removed a ton of the props being passed from the thread => messages component, mostly because they were unnecessary or were causing misdirection in understanding where logic was being calculated for things like the force scrolling - most of that logic now lives in the messages `componentDidUpdate` method, versus being computed in the thread container and being passed as props.

File: src/views/thread/components/messages.js [hunk 2]
```diff
@@ -147,33 +146,55 @@ class MessagesWithData extends React.Component<Props, State> {
...
 
       const uniqueMessages = deduplicateChildren(unsortedMessages, 'id');
       const sortedMessages = sortAndGroupMessages(uniqueMessages);
 
+      const prevCursor =
+        edges && edges.length > 0 && edges[0] && edges[0].cursor;
+      const lastEdge =
+        edges &&
+        edges.length > 0 &&
+        edges[edges.length - 1] !== (null || undefined) &&
```
Comment (mxstbr) [inline src/views/thread/components/messages.js]: This don't work, `(null || undefined)` evaluates to false. This check is unnecessary anyway, just make it return `edges[edges.length - 1]`!

**Review by mxstbr — DISMISSED** [review]
This is looking super great apart from one nitpick, if it fixes the prob for you let's get this on alpha!

**Pull request merged**